bench/src/bulk.rs

```diff
@@ -80,34 +136,143 @@ async fn client(server_addr: SocketAddr, server_cert: quinn::Certificate) -> Res
+        send_data_on_stream(get_connection(), config.stream_size_mb).await
+    }))
+    .buffer_unordered(config.max_streams);
+
+    let mut total_size = 0;
+    let mut duration_hist: Histogram<u64> = Histogram::new(3).unwrap();
+    let mut throughput_hist: Histogram<u64> = Histogram::new(3).unwrap();
+
+    while let Some(result) = ops.next().await {
+        info!("stream finished: {:?}", result);
+        let result = result.unwrap();
+        total_size += result.size;
+
+        duration_hist
+            .record(result.duration.as_millis() as u64)
+            .unwrap();
+        throughput_hist.record(result.throughput as u64).unwrap();
     }
-    stream.finish().await.context("failed finishing stream")?;
-    println!("sent {} bytes in {:?}", 1024 * DATA.len(), start.elapsed());
+
+    let dt = start.elapsed();
+    println!("Overall stats:\n");
+    println!(
+        "Sent {} bytes on {} streams in {:4.2?} ({:.2} MiB/s)\n",
+        total_size,
+        config.streams,
+        dt,
+        throughput_bps(dt, total_size as u64) / 1024.0 / 1024.0
+    );
+
+    println!("Stream metrics:\n");
+
+    println!(
+        "AVG : Throughput: {:7.2} MiB/s, Duration: {:.2?}",
+        throughput_hist.mean() as f64 / 1024.0 / 1024.0,
+        Duration::from_millis(duration_hist.mean() as u64)
+    );
+    println!(
+        "P0  : Throughput: {:7.2} MiB/s, Duration: {:.2?}",
+        throughput_hist.value_at_quantile(0.00) as f64 / 1024.0 / 1024.0,
+        Duration::from_millis(duration_hist.value_at_quantile(0.00))
+    );
+    println!(
+        "P10 : Throughput: {:7.2} MiB/s, Duration: {:.2?}",
+        throughput_hist.value_at_quantile(0.00) as f64 / 1024.0 / 1024.0,
+        Duration::from_millis(duration_hist.value_at_quantile(0.10))
+    );
+    println!(
+        "P50 : Throughput: {:7.2} MiB/s, Duration: {:.2?}",
+        throughput_hist.value_at_quantile(0.50) as f64 / 1024.0 / 1024.0,
+        Duration::from_millis(duration_hist.value_at_quantile(0.50))
+    );
+    println!(
+        "P90 : Throughput: {:7.2} MiB/s, Duration: {:.2?}",
+        throughput_hist.value_at_quantile(0.99) as f64 / 1024.0 / 1024.0,
+        Duration::from_millis(duration_hist.value_at_quantile(0.90))
+    );
+    println!(
+        "P100: Throughput: {:7.2} MiB/s, Duration: {:.2?}",
+        throughput_hist.value_at_quantile(1.00) as f64 / 1024.0 / 1024.0,
+        Duration::from_millis(duration_hist.value_at_quantile(1.00))
+    );
+
+    // Explicit close of the connection, since handles can still be around due
+    // to `Arc`ing them
+    connection.close(0u32.into(), b"Benchmark done");
+
+    endpoint.wait_idle().await;
+
     Ok(())
 }
 
+fn throughput_bps(duration: Duration, size: u64) -> f64 {
+    (size as f64) / (duration.as_secs_f64())
+}
+
 fn rt() -> Runtime {
     Builder::new()
         .basic_scheduler()
         .enable_all()
         .build()
         .unwrap()
 }
+
+fn transport_config() -> quinn::TransportConfig {
```

Comment:
I think I'd prefer to keep this inline and just clone the `transport_config`, what do you think?

Reply:
It's created 2 times - once for the server and once for the client in 2 different functions. I think when trying to play around with new values it's easiest to modify both in one place. Therefore I moved it out (and actually there wasn't any explicit configuration before).

Reply:
Sorry, must have missed that -- I did find the two usages but thought they were in the same function.